homeassistant/components/intellifire/fan.py

```diff
@@ -0,0 +1,142 @@
+    value_fn: Callable[[IntellifirePollData], bool]
+    data_field: str
+    named_speeds: list[str]
+
+
+@dataclass
+class IntellifireFanEntityDescription(
+    FanEntityDescription, IntellifireFanRequiredKeysMixin
+):
+    """Describes a fan entity."""
+
+
+INTELLIFIRE_FANS: tuple[IntellifireFanEntityDescription, ...] = (
+    IntellifireFanEntityDescription(
+        key="fan",
+        name="Fan",
+        set_fn=lambda control_api, speed: control_api.set_fan_speed(speed=speed),
+        value_fn=lambda data: data.fanspeed,
+        data_field="fanspeed",
+        named_speeds=[
+            "quiet",
+            "low",
+            "medium",
+            "high",
+        ],  # off is not included
+    ),
+)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up the fans."""
+    coordinator: IntellifireDataUpdateCoordinator = hass.data[DOMAIN][entry.entry_id]
+
+    if coordinator.data.has_fan:
+        async_add_entities(
+            IntellifireFan(coordinator=coordinator, description=description)
+            for description in INTELLIFIRE_FANS
+        )
+        return
+    LOGGER.info("Disabling Fan - IntelliFire device does not appear to have one")
+
+
+class IntellifireFan(IntellifireEntity, FanEntity):
+    """This is Fan entity for the fireplace."""
+
+    entity_description: IntellifireFanEntityDescription
+    _attr_supported_features = FanEntityFeature.SET_SPEED | FanEntityFeature.PRESET_MODE
+
+    @property
+    def is_on(self) -> bool:
+        """Return on or off."""
+        return self.entity_description.value_fn(self.coordinator.read_api.data) >= 1
+
+    @property
+    def percentage(self) -> int | None:
+        """Return fan percentage."""
+        percent_step = ordered_list_item_to_percentage(
+            self.entity_description.named_speeds,
+            self.entity_description.named_speeds[0],
+        )
+        return (
+            self.entity_description.value_fn(self.coordinator.read_api.data)
+            * percent_step
+        )
+
+    @property
+    def speed_count(self) -> int:
+        """Count of supported speeds."""
+        return len(self.entity_description.named_speeds)
+
+    async def async_set_preset_mode(self, preset_mode: str) -> None:
+        """Set the preset mode of the fan."""
+        # Get a percent from the preset
+
+        percent = ordered_list_item_to_percentage(
+            self.entity_description.named_speeds, preset_mode
+        )
+        await self.async_set_percentage(percent)
+
+    async def async_set_percentage(self, percentage: int) -> None:
+        """Set the speed percentage of the fan."""
+        # Calculate percentage steps
+        LOGGER.debug("Setting Fan Speed %s", percentage)
+        percent_step = 100.0 / len(self.entity_description.named_speeds)
+        int_value = int(math.ceil(float(percentage) / percent_step))
```

Comment:
The fan only has 4 possible speeds:

What I've done is taken the 1,2,3,4 and turned them into quiet, low, medium, high ... it can't actually accept a percentage. ....

Is this an incorrect approach that I've taken?


<img width="1084" alt="image" src="https://user-images.githubusercontent.com/6491743/179231024-48c9def4-5c6a-48e4-bb3a-fa1d4133f505.png">

Reply:
You could convert 1 -> 25%, 2->50%, 3->75%, 4->100% with the percentage to ranged value helper